posthog/async_migrations/migrations/0006_persons_and_groups_on_events_backfill.py

```diff
@@ -0,0 +1,259 @@
+from functools import cached_property
+
+import structlog
+
+from posthog.async_migrations.definition import (
+    AsyncMigrationDefinition,
+    AsyncMigrationOperation,
+    AsyncMigrationOperationSQL,
+    AsyncMigrationType,
+)
+from posthog.async_migrations.utils import execute_on_all_shards_in_parallel, run_optimize_table
+from posthog.client import sync_execute
+from posthog.models.event.sql import EVENTS_DATA_TABLE
+from posthog.settings import CLICKHOUSE_DATABASE
+
+logger = structlog.get_logger(__name__)
+
+"""
+Migration summary
+=================
+
+Backfill the sharded_events table to add data for the following columns:
+
+- person_id
+- person_properties
+- person_created_at
+- group_X_properties
+- group_X_created_at
+
+This allows us to switch entirely to only querying the events table for insights,
+without having to hit additional tables for persons, groups, and distinct IDs.
+
+Migration strategy
+==================
+
+We will run the following operations on the cluster
+(or on one node per shard if shard-level configuration is provided):
+
+1. Create temporary tables with the relevant columns from `person`, `person_distinct_id`, and `groups`
+2. Insert data from the main tables into them
+3. Optimize the temporary tables to remove duplicates
+4. Create a dictionary to query each temporary table
+5. Run an ALTER TABLE ... UPDATE to backfill all the data using the dictionaries
+"""
+
+TEMPORARY_PERSONS_TABLE_NAME = "tmp_person_0006"
+TEMPORARY_PDI2_TABLE_NAME = "tmp_person_distinct_id2_0006"
+TEMPORARY_GROUPS_TABLE_NAME = "tmp_groups_0006"
+
+
+class Migration(AsyncMigrationDefinition):
+    description = "Backfill persons and groups data on the sharded_events table"
+
+    depends_on = "0005_person_replacing_by_version"
+
+    def precheck(self):
+        return True, None
+
+    def is_required(self) -> bool:
+        rows = sync_execute(
+            """
+            SELECT comment
+            FROM system.columns
+            WHERE database = %(database)s AND
+            table = %(events_data_table)s
+        """,
+            {"database": CLICKHOUSE_DATABASE, "events_data_table": EVENTS_DATA_TABLE()},
+        )
+
+        comments = [row[0] for row in rows]
+        return "skip_0006_persons_and_groups_on_events_backfill" not in comments
+
+    def run_backfill(self, _):
```

Comment:
Nit: Ordering. We've always ordered async migration functions as is_required, operations, rest.

This should be _after_ def operations.